### PR TITLE
[For Review] CA-225070-v2: Refined order of xapi/xhad/attach-static-vdis start/stop

### PR DESCRIPTION
**Pull request by zli**
This is an alternative version of #2827.

It's slightly better as it doesn't hard wire the xhad (the default cluster-stack) commands. But it comes with the cost of adding an optional parameter to an existing API method.

### CODE REVIEW

File: ocaml/xapi/xapi_ha.ml [hunk 1]
```diff
@@ -834,27 +834,35 @@ let ha_stop_daemon __context localhost =
...
-  begin
-    try
-      ha_stop_daemon __context ();
-    with Xha_error e ->
-      error "Host.emergency_ha_disable: ha_stop_daemon failed with %s; continuing" (Xha_errno.to_string e)
-  end;
-  (* Might not be able to access the database to detach statefiles; however this isn't critical *)
-  ()
+let emergency_ha_disable __context soft =
+  let ha_armed = Localdb.get Constants.ha_armed in
```
Comment (lindig) [inline ocaml/xapi/xapi_ha.ml]: Minor suggestion: `let ha_armed = Localdb.get Constants.ha_armed = "armed"`

File: ocaml/idl/datamodel.ml [hunk 2]
```diff
@@ -4475,7 +4475,9 @@ let host_emergency_ha_disable = call ~flags:[`Session]
     ~name:"emergency_ha_disable"
     ~in_oss_since:None
     ~in_product_since:rel_orlando
-    ~params:[]
+    ~versioned_params:
+      [{param_type=Bool; param_name="soft"; param_doc="Disable temporarily, revert upon host reboot or further changes, idempotent"; param_release=dundee_plus_release; param_default=Some(VBool false)};
```
Comment (lindig) [inline ocaml/idl/datamodel.ml]: I'd say `Disable HA temporarily, revert ...`.

Comment (zli) [inline ocaml/idl/datamodel.ml]: Updated.

**Comment by mseri** [comment]
We are waiting for the branching to complete so that we can merge this PR and run tests

File: ocaml/xapi/xapi_ha.ml [hunk 3]
```diff
@@ -834,27 +834,35 @@ let ha_stop_daemon __context localhost =
...
+  if ha_armed = "false" then
+    if soft then () else raise (Api_errors.Server_error(Api_errors.ha_not_enabled, []))
+  else begin
+    warn "Host.emergency_ha_disable: Disabling the HA subsystem on the local host only.";
+    Localdb.put Constants.ha_armed "false";
```
Comment (jonludlam) [inline ocaml/xapi/xapi_ha.ml]: What's the reasoning behind setting this here and unsetting later in the 'soft=true' case?

**Comment by jonludlam** [comment]
@zli - short question to answer above, then we're good to merge this.

File: ocaml/xapi/xapi_ha.ml (hunk 3 above)
Comment (zli) [inline ocaml/xapi/xapi_ha.ml]: No particular reason except that the code been replaced also changed the config before the disabling action.

Nonetheless, after checking the places referencing this config variable, it seem that changing the config afterwards could be more conservative/safer (if there is any difference). 

As such, I've updated the code accordingly.

**Pull request merged**